sui_core/src/safe_client.rs

```diff
@@ -172,6 +170,44 @@ impl<C> SafeClient<C> {
         Ok(())
     }
 
+    fn check_update_item_batch_response(
+        &self,
+        request: BatchInfoRequest,
+        signed_batch: &SignedBatch,
+    ) -> SuiResult {
+        // check the signature of the batch
+        signed_batch
+            .signature
+            .check(&signed_batch.batch, signed_batch.authority)?;
+
+        // ensure transactions enclosed match requested range
+        fp_ensure!(
+            signed_batch.batch.initial_sequence_number >= request.start
+                && signed_batch.batch.next_sequence_number
+                    <= (request.end + signed_batch.batch.size),
+            SuiError::ByzantineAuthoritySuspicion {
+                authority: self.address
+            }
+        );
+
+        // reconstruct the batch and make sure the constructed digest matches the provided one
```

Comment:
This section of code checks that the transactions digests match (which are included in the batch). But instead it should check that the digests of / or the batches themselves match (ie variables `signed_batch.batch` and `reconstructed_batch`). This ensures that all information expected matches the information contained in the signature.

But in this case I do not think they will: 
```
let reconstructed_batch = AuthorityBatch::make_next_with_previous_digest(
            Some(provided_digest),
            &signed_batch.batch.transaction_batch.0,
        );
```
In the above the `Some(provided_digest)` is provided instead of the digest of the previous batch. Also, the input to this function should include the previous batch digest if any (to reconstruct the expected batch) as well as the list of transaction seq / digests received to ensure they are the ones fed into the constructor `AuthorityBatch::make_next_with_previous_digest`, again to ensure that what has been received in terms of sequence of transactions is actually what is eventually signed.